File: java/416_PartitionEqualSubsetSum.java
class Solution {
    HashMap<Integer, Boolean> memo = new HashMap<Integer, Boolean>();
    public boolean canPartition(int[] nums) {
        int N = nums.length;
        int sum = 0;
        for (int n : nums) sum += n;
        return sum % 2 == 0 & helper(nums, N - 1, sum / 2);
    }
    
    public boolean helper(int[] nums, int idx, int currentTargetSum) {
        if (memo.containsKey(currentTargetSum)) return memo.get(currentTargetSum);
        else if (currentTargetSum == 0) return true;
        else if (idx == 0 || currentTargetSum < 0) return false;
        
        memo.put(currentTargetSum, helper(nums, idx - 1, currentTargetSum - nums[idx]) || helper(nums, idx - 1, currentTargetSum));
        return memo.get(currentTargetSum);
    }
}

File: python/416_PartitionEqualSubsetSum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        def dfs(idx, curr_target):
            if curr_target in memo:
                return memo[curr_target]
            elif curr_target == 0:
                return True
            elif idx == 0 or curr_target < 0:
                return False
            
            memo[curr_target] = dfs(idx - 1, curr_target - nums[idx]) or dfs(idx - 1, curr_target)
            return memo[curr_target]
            
        memo = {}
        total = sum(nums)
        return total % 2 == 0 and dfs(len(nums) - 1, total // 2)
        
